slow_conv2d_forward: avoid calling dispatcher in parallel region

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65759
* #65758
* #65757
* #65737
* #65726
* #65725
* __->__ #65724

See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Copy bias into output only once, outside of the parallel region
3. Replaces `addmm`_ with a direct call to gemm.

Technically this also adds a new requirement that the output always be
contiguous, but the out argument version isn't exposed or used
anywhere in the `torch.nn` API. So that should be fine.

Differential Revision: [D31257875](https://our.internmc.facebook.com/intern/diff/D31257875)

peterbell10 committed:
slow_conv2d_forward: avoid calling dispatcher in parallel region

See gh-56794

This replaces Tensor slicing with TensorAccessor slicing,
and replaces addmm_ with a direct call to gemm_.

Technically this adds a new requirement that the output always be
contiguous, but the out argument version isn't exposed or used
anywhere in the `torch.nn` API. So that should be fine.

[ghstack-poisoned]

peterbell10 committed:
Update on "slow_conv2d_forward: avoid calling dispatcher in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Copy bias into output only once, outside of the parallel region
3. Replaces `addmm`_ with a direct call to gemm.

Technically this also adds a new requirement that the output always be
contiguous, but the out argument version isn't exposed or used
anywhere in the `torch.nn` API. So that should be fine.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -410,34 +417,44 @@ std::tuple<Tensor&, Tensor&> slow_conv2d_forward_out_cpu(
+  if (bias.defined()) {
+    output.copy_(bias.reshape({-1, 1, 1}));
+  }
+  TORCH_CHECK(output.is_contiguous());
```
can you also please check that `finput` is contiguous? It wasn't done before, but `finput` can be discontiguous if it is passed as a discontiguous tensor of the correct size.

peterbell10 committed:
Update on "slow_conv2d_forward: avoid calling dispatcher in parallel region"


See gh-56794

Avoid dispatch inside of parallel_for by:
1. Replacing Tensor slicing with TensorAccessor
2. Copy bias into output only once, outside of the parallel region
3. Replaces `addmm`_ with a direct call to gemm.

Technically this also adds a new requirement that the output always be
contiguous, but the out argument version isn't exposed or used
anywhere in the `torch.nn` API. So that should be fine.

[ghstack-poisoned]

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31257875).

ngimel commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -410,34 +417,45 @@ std::tuple<Tensor&, Tensor&> slow_conv2d_forward_out_cpu(
+  TORCH_CHECK(output.is_contiguous() && finput.is_contiguous(),
+              "slow_conv2d output tensors must be contiguous");
+
+  AT_DISPATCH_ALL_TYPES_AND(kBFloat16, input.scalar_type(), "slow_conv2d_cpu", [&]{
```
I don't think ALL_TYPES and bfloat16 makes sense here. In contrast to addmm, fast bfloat16 support is not enabled on direct calls to at::cpublas::gemm, and in CPUBlas.cpp there are specializations only for float, double, complex float and double and int64, the rest go to extremely slow `gemm_stub`.

peterbell10 replied:
These are the dtypes that conv2d supports. As limited by Unfold2d, which is only used in this code:
https://github.com/pytorch/pytorch/blob/1d681c1ab2e335de2ec32995dc2aa4e57dd873e7/aten/src/ATen/native/cpu/Unfold2d.cpp#L287-L288

Although this did make me realize that supporting `ALL_TYPES` for backward functions doesn't make any sense. So I've changed those PRs to use `FLOATING_TYPES_AND(kBFloat16`

>In contrast to addmm, fast bfloat16 support is not enabled on direct calls to at::cpublas::gemm

I assume this is #64619. That seems like a weird choice to me. Do you know why they special case addmm instead of gemm?

ngimel replied:
Probably because it allowed them to have a single `matmul` function instead of separate `gemm` and batched gemm, also iirc their `matmul` can handle broadcasting in the addition part, that gemm cannot, and integrating at `gemm` level would not have allowed them to take advantage of that (if `b` in `ax+b` has to be broadcasted, we first explicitly expand it and them call gemm)

ngimel commented:
This stack is causing mobile failures, so will take some time
